src/routes/api/client-credentials/microsoft/token.ts

```diff
@@ -0,0 +1,163 @@
+
+  class ResponseDTO implements ToJSON {
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.expires_in, { toClassOnly: true })
+    public readonly expiresIn!: number;
+
+    @Expose()
+    @Transform(({ value }) => new Date(value).toUTCString(), { toPlainOnly: true })
+    public readonly expiresAt?: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.token_type, { toClassOnly: true })
+    public readonly tokenType!: string;
+
+    @Expose()
+    @Transform(({ obj, value }) => value ?? obj.access_token, { toClassOnly: true })
+    public readonly accessToken!: string;
+
+    toJSON(): Exclude<JSONValue, ToJSON> {
+      return instanceToPlain(this);
+    }
+  }
+
+  // validate request query
+  const searchParams = Object.fromEntries(requestEvent.url.searchParams);
+  const query = plainToInstance(RequestQueryDTO, searchParams, classTransformOptions);
+  const queryValidationErrors = validateSync(query, classValidationOptions);
+  if (queryValidationErrors.length > 0) {
+    return { status: 400, body: { query: queryValidationErrors } };
+  }
+
+  // validate request header
+  const obj = humps.camelizeKeys(Object.fromEntries(requestEvent.request.headers.entries()));
+  const header = plainToInstance(RequestHeaderDTO, obj, classTransformOptions);
+  const headerValidationErrors = validateSync(header, classValidationOptions);
+  if (headerValidationErrors.length > 0) {
+    return { status: 401 };
+  }
+
+  const orm = await MikroORM.init({ ...config, ...{ entities: [Activation] } });
+  const em = orm.em.fork();
+
+  const subject = webexHttpPeopleResource(header.authorization)
+    .getMyOwnDetails()
+    .then((r) => r.json())
+    .then((r) => r.emails[0]);
+
+  const account = em.findOneOrFail(Activation, query.activationId, {
+    fields: ['microsoftTenantId', 'microsoftClientId', 'microsoftPrivateKey', 'microsoftClientCertificate']
+  });
+
+  const formJwt = (audience: string, subject: string, issuer: string, pkcs8: string, x509: string, expiry = '1h') => {
+    const alg = 'RS256';
+    const data = x509
+      .replaceAll('\n', '')
+      ?.match(/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/gm)
+      ?.join('');
+    const sha1 = crypto.subtle.digest('SHA-1', Buffer.from(data as string, 'base64'));
+    const hex = sha1.then((r) => Buffer.from(r).toString('hex'));
+    const x5t = hex.then((r) => Buffer.from(r, 'hex').toString('base64url'));
+
+    return Promise.all([x5t, jose.importPKCS8(pkcs8, alg)]).then(([r1, r2]) =>
+      new jose.SignJWT({})
+        .setProtectedHeader({ alg, x5t: r1 })
+        .setIssuer(issuer)
+        .setAudience(audience)
+        .setSubject(subject)
+        .setIssuedAt()
+        .setExpirationTime(expiry)
+        .sign(r2)
+    );
+  };
+
+  const addExpiresAt = (obj: JSONObject, date: number = Date.now()) => ({
+    ...obj,
+    ...{ expiresAt: date + (obj.expiresIn as number) * 1000 }
+  });
+
+  return Promise.all([subject, account])
```

Comment:
in couple of sentences can you explain what this promise resolve? nested promises are kinda hard to follow IMO.

Reply:
Essentially generates a JWT. The flow is described here: https://stackoverflow.com/a/75168441.